File: kotlin/src/main/kotlin/foundation/zoobc/zbc/Body.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
package foundation.zoobc.zbc

/** Parameter validation and the generic body serialiser driven by spec/transactions (encodings of index.json). */
object Body {
    @JvmStatic fun parseInteger(value: String, kind: String, name: String): Long {
        val v = value.trim()
        if (!Regex("^-?\\d+$").matches(v)) throw ToolError.usage("$name must be a whole number, got \"$value\"")
        val n = v.toBigIntegerOrNull() ?: throw ToolError.usage("$name is out of range for $kind")
        val (lo, hi) = when (kind) {
            "uint64" -> java.math.BigInteger.ZERO to java.math.BigInteger("18446744073709551615")
            "uint32" -> java.math.BigInteger.ZERO to java.math.BigInteger.valueOf(0xffffffffL)
            "uint8" -> java.math.BigInteger.ZERO to java.math.BigInteger.valueOf(255)
            else -> java.math.BigInteger.valueOf(Long.MIN_VALUE) to java.math.BigInteger.valueOf(Long.MAX_VALUE)
        }
        if (n < lo || n > hi) throw ToolError.usage("$name is out of range for $kind")
        return n.toLong()
    }

    @JvmStatic fun splitList(s: String): List<String> = s.split(",").map { it.trim() }.filter { it.isNotEmpty() }

    /** Check one value against its parameter kind; returns the value to keep. */
    @JvmStatic fun validateParam(p: ParamDef, value: String): String {
        when (p.kind) {
            "privkey" -> if (!Encoding.isHex(value, 64)) throw ToolError.usage("${p.name} must be 64 hex characters (a 32-byte private key)")
            "address" -> try { Address.parse(value) } catch (e: IllegalArgumentException) { throw ToolError.usage("invalid ${p.name}: ${e.message}") }
            "address_list" -> for (a in splitList(value)) try { Address.parse(a) } catch (e: IllegalArgumentException) { throw ToolError.usage("invalid ${p.name} entry $a: ${e.message}") }
            "key" -> try { Address.parseKey32(value) } catch (e: IllegalArgumentException) { throw ToolError.usage("invalid ${p.name}: ${e.message}") }
            "int64", "uint64", "uint32", "uint8" -> {
                val n = parseInteger(value, p.kind, p.name)
                if ((p.min != null && n < p.min) || (p.max != null && n > p.max)) throw ToolError.usage("${p.name} must be between ${p.min ?: "-inf"} and ${p.max ?: "inf"}")
                return value.trim()
            }
            "hex32" -> if (!Encoding.isHex(value, 64)) throw ToolError.usage("${p.name} must be 64 hex characters (32 bytes)")
            "hexbytes" -> if (!Encoding.isHex(value)) throw ToolError.usage("${p.name} must be hex")
        }
        return value
    }

    /** What the generic serialiser cannot read from the parameters. */
    class Context(val sender: KeyPair? = null, val files: Map<String, ByteArray> = emptyMap(), val computed: MutableMap<String, ByteArray> = mutableMapOf())

    private fun holds(whenExpr: String, params: Map<String, String>): Boolean {
        val m = Regex("^(\\w+) != (0|'')$").matchEntire(whenExpr) ?: throw ToolError.internal("unsupported condition $whenExpr")
        val v = params[m.groupValues[1]] ?: ""
        return if (m.groupValues[2] == "0") v.isNotEmpty() && (v.toLongOrNull() ?: 0L) != 0L else v.isNotEmpty()
    }

    /** Serialise one body field. */
    @JvmStatic fun encodeField(f: FieldDef, params: Map<String, String>, ctx: Context): ByteArray {
        ctx.computed[f.name]?.let { return it }
        var value = params[f.from] ?: ""
        if (f.when_zero != null && (value.isEmpty() || (value.toLongOrNull() ?: 0L) <= 0L)) value = params[f.when_zero] ?: "0"
        return when (f.encoding) {
            "u8" -> byteArrayOf(parseInteger(value, "uint8", f.name).toByte())
            "u16le" -> Bytes.le16(parseInteger(value, "uint32", f.name).toInt())
            "u32le" -> Bytes.le32(parseInteger(value, "uint32", f.name).toInt())
            "u64le" -> Bytes.le64(parseInteger(value, "int64", f.name))
            "hex" -> { val b = hexOrUsage(value, f.from); if (f.size != null && b.size != f.size) throw ToolError.usage("${f.from} must be ${f.size} bytes (${2 * f.size} hex)"); b }
            "hex16" -> { val b = hexOrUsage(value, f.from); Bytes.le16(b.size) + b }
            "bytes32" -> { val b = ctx.files[f.from] ?: hexOrUsage(value, f.from); Bytes.le32(b.size) + b }
            "str16" -> { val b = value.toByteArray(); Bytes.le16(b.size) + b }
            "str32" -> { val b = value.toByteArray(); Bytes.le32(b.size) + b }
            "address" -> addr(value, f.from).bytes
            "address_list" -> splitList(value).fold(ByteArray(0)) { acc, a -> acc + addr(a, f.from).bytes }
            "address_list8" -> { val items = splitList(value); if (items.size > 255) throw ToolError.usage("${f.from}: at most 255 entries"); items.fold(byteArrayOf(items.size.toByte())) { acc, a -> acc + addr(a, f.from).bytes } }
            "sender_address" -> (ctx.sender ?: throw ToolError.internal("no sender")).accountBytes
            "pubkey_of_key" -> try { KeyPair.fromHex(value).publicKey } catch (e: IllegalArgumentException) { throw ToolError.usage("${f.from} must be 64 hex characters (a 32-byte private key)") }
            "key32" -> try { Address.parseKey32(value) } catch (e: IllegalArgumentException) { throw ToolError.usage("invalid ${f.from}: ${e.message}") }
            "literal" -> Encoding.hexToBytes(f.value ?: "")
            "custom" -> throw ToolError.internal("field ${f.name} needs a custom hook")
            else -> throw ToolError.internal("unknown encoding ${f.encoding}")
        }
    }

    private fun hexOrUsage(v: String, name: String) = if (Encoding.isHex(v)) Encoding.hexToBytes(v) else throw ToolError.usage("$name must be hex")
    private fun addr(v: String, name: String) = try { Address.parse(v) } catch (e: IllegalArgumentException) { throw ToolError.usage("invalid $name: ${e.message}") }

    /** The whole body of a non-custom transaction. */
    @JvmStatic fun build(def: TxDef, params: Map<String, String>, ctx: Context): ByteArray {
        var out = ByteArray(0)
        for (f in def.body) {
            if (f.`when` != null && !holds(f.`when`, params)) continue
            out += encodeField(f, params, ctx)
        }
        return out
    }
}

File: swift/Sources/ZBC/Body.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci

import Foundation

/// Parameter validation and the generic body serialiser driven by spec/transactions (encodings of index.json).
public enum Body {
    public static func parseInteger(_ value: String, kind: String, name: String) throws -> Int64 {
        let v = value.trimmingCharacters(in: .whitespaces)
        let digits = v.hasPrefix("-") ? String(v.dropFirst()) : v
        guard !digits.isEmpty, digits.allSatisfy({ $0.isASCII && $0.isNumber }) else { throw ToolError.usage("\(name) must be a whole number, got \"\(value)\"") }
        switch kind {
        case "uint64": guard let u = UInt64(v) else { throw ToolError.usage("\(name) is out of range for \(kind)") }; return Int64(bitPattern: u)
        case "uint32": guard let u = UInt32(v) else { throw ToolError.usage("\(name) is out of range for \(kind)") }; return Int64(u)
        case "uint8": guard let u = UInt8(v) else { throw ToolError.usage("\(name) is out of range for \(kind)") }; return Int64(u)
        default: guard let n = Int64(v) else { throw ToolError.usage("\(name) is out of range for \(kind)") }; return n
        }
    }

    public static func splitList(_ s: String) -> [String] { s.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty } }

    /// Check one value against its parameter kind; returns the value to keep.
    public static func validateParam(_ p: ParamDef, _ value: String) throws -> String {
        switch p.kind {
        case "privkey": guard Enc.isHex(value, 64) else { throw ToolError.usage("\(p.name) must be 64 hex characters (a 32-byte private key)") }
        case "address": do { _ = try Address.parse(value) } catch let e as Address.Invalid { throw ToolError.usage("invalid \(p.name): \(e.message)") }
        case "address_list": for a in splitList(value) { do { _ = try Address.parse(a) } catch let e as Address.Invalid { throw ToolError.usage("invalid \(p.name) entry \(a): \(e.message)") } }
        case "key": do { _ = try Address.parseKey32(value) } catch let e as Address.Invalid { throw ToolError.usage("invalid \(p.name): \(e.message)") }
        case "int64", "uint64", "uint32", "uint8":
            let n = try parseInteger(value, kind: p.kind, name: p.name)
            if let lo = p.min, n < lo { throw ToolError.usage("\(p.name) must be between \(lo) and \(p.max.map(String.init) ?? "inf")") }
            if let hi = p.max, n > hi { throw ToolError.usage("\(p.name) must be between \(p.min.map(String.init) ?? "-inf") and \(hi)") }
            return value.trimmingCharacters(in: .whitespaces)
        case "hex32": guard Enc.isHex(value, 64) else { throw ToolError.usage("\(p.name) must be 64 hex characters (32 bytes)") }
        case "hexbytes": guard Enc.isHex(value) else { throw ToolError.usage("\(p.name) must be hex") }
        default: break
        }
        return value
    }

    /// What the generic serialiser cannot read from the parameters.
    public final class Context {
        public var sender: KeyPair?
        public var files: [String: [UInt8]]
        public var computed: [String: [UInt8]]
        public init(sender: KeyPair? = nil, files: [String: [UInt8]] = [:], computed: [String: [UInt8]] = [:]) { self.sender = sender; self.files = files; self.computed = computed }
    }

    private static func holds(_ when: String, _ params: [String: String]) throws -> Bool {
        let parts = when.components(separatedBy: " != ")
        guard parts.count == 2 else { throw ToolError.internalError("unsupported condition \(when)") }
        let v = params[parts[0]] ?? ""
        if parts[1] == "0" { return !v.isEmpty && (Int64(v) ?? 0) != 0 }
        if parts[1] == "''" { return !v.isEmpty }
        throw ToolError.internalError("unsupported condition \(when)")
    }

    private static func hexOrUsage(_ v: String, _ name: String) throws -> [UInt8] { guard let b = Enc.unhex(v) else { throw ToolError.usage("\(name) must be hex") }; return b }
    private static func addr(_ v: String, _ name: String) throws -> ParsedAddress { do { return try Address.parse(v) } catch let e as Address.Invalid { throw ToolError.usage("invalid \(name): \(e.message)") } }

    /// Serialise one body field.
    public static func encodeField(_ f: FieldDef, _ params: [String: String], _ ctx: Context) throws -> [UInt8] {
        if let v = ctx.computed[f.name] { return v }
        var value = params[f.from] ?? ""
        if let wz = f.when_zero, value.isEmpty || (Int64(value) ?? 0) <= 0 { value = params[wz] ?? "0" }
        switch f.encoding {
        case "u8": return [UInt8(try parseInteger(value, kind: "uint8", name: f.name))]
        case "u16le": return LE.u16(Int(try parseInteger(value, kind: "uint32", name: f.name)))
        case "u32le": return LE.u32(UInt32(try parseInteger(value, kind: "uint32", name: f.name)))
        case "u64le": return LE.i64(try parseInteger(value, kind: "int64", name: f.name))
        case "hex":
            let b = try hexOrUsage(value, f.from)
            if let size = f.size, b.count != size { throw ToolError.usage("\(f.from) must be \(size) bytes (\(2 * size) hex)") }
            return b
        case "hex16": let b = try hexOrUsage(value, f.from); return LE.u16(b.count) + b
        case "bytes32": let b = try ctx.files[f.from] ?? hexOrUsage(value, f.from); return LE.u32(UInt32(b.count)) + b
        case "str16": let b = Array(value.utf8); return LE.u16(b.count) + b
        case "str32": let b = Array(value.utf8); return LE.u32(UInt32(b.count)) + b
        case "address": return try addr(value, f.from).bytes
        case "address_list": return try splitList(value).flatMap { try addr($0, f.from).bytes }
        case "address_list8":
            let items = splitList(value)
            if items.count > 255 { throw ToolError.usage("\(f.from): at most 255 entries") }
            return [UInt8(items.count)] + (try items.flatMap { try addr($0, f.from).bytes })
        case "sender_address": guard let s = ctx.sender else { throw ToolError.internalError("no sender") }; return s.accountBytes
        case "pubkey_of_key": guard let kp = try? KeyPair(hex: value) else { throw ToolError.usage("\(f.from) must be 64 hex characters (a 32-byte private key)") }; return kp.publicKey
        case "key32": do { return try Address.parseKey32(value) } catch let e as Address.Invalid { throw ToolError.usage("invalid \(f.from): \(e.message)") }
        case "literal": return Enc.unhex(f.value ?? "") ?? []
        case "custom": throw ToolError.internalError("field \(f.name) needs a custom hook")
        default: throw ToolError.internalError("unknown encoding \(f.encoding)")
        }
    }

    /// The whole body of a non-custom transaction.
    public static func build(_ def: TxDef, _ params: [String: String], _ ctx: Context) throws -> [UInt8] {
        var out: [UInt8] = []
        for f in def.body {
            if let w = f.when, !(try holds(w, params)) { continue }
            out += try encodeField(f, params, ctx)
        }
        return out
    }
}
